backend/asm_targets/asm_directives.ml

```diff
@@ -0,0 +1,322 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER         *
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,  *
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE  *
+ * SOFTWARE.                                                                      *
+ *                                                                                *
+ **********************************************************************************)
+
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+module type S = Asm_directives_intf.S
+
+module type Arg = Asm_directives_intf.Arg
+
+module Cached_string = struct
+  type t =
+    { section : Asm_section.t;
+      str : string;
+      comment : string option
+    }
+
+  include Identifiable.Make (struct
+    type nonrec t = t
+
+    let compare { section = section1; str = str1; comment = comment1 }
+        { section = section2; str = str2; comment = comment2 } =
+      let c = Asm_section.compare section1 section2 in
+      if c <> 0
+      then c
+      else
+        let c = String.compare str1 str2 in
+        if c <> 0 then c else Option.compare String.compare comment1 comment2
+
+    let equal t1 t2 = compare t1 t2 = 0
+
+    let hash t = Hashtbl.hash t
+
+    let print _ _ = Misc.fatal_error "Not yet implemented"
+
+    let output _ _ = Misc.fatal_error "Not yet implemented"
+  end)
+end
+
+module Make (A : Asm_directives_intf.Arg) : Asm_directives_intf.S = struct
+  module Int8 = Numbers.Int8
+  module Int16 = Numbers.Int16
+  module Uint8 = Numbers.Uint8
+  module Uint16 = Numbers.Uint16
+  module Uint32 = Numbers.Uint32
+  module Uint64 = Numbers.Uint64
+  module D = A.D
+
+  let sections_seen = ref []
+
+  let current_section_ref = ref None
+
+  let cached_strings = ref Cached_string.Map.empty
+
+  let temp_var_counter = ref 0
+
+  let is_macos () =
+    match Target_system.assembler () with
+    | MASM | GAS_like -> false
+    | MacOS -> true
+
+  let if_not_masm f =
+    match Target_system.assembler () with
+    | MASM -> ()
+    | GAS_like | MacOS -> f ()
+
+  let loc ~file_num ~line ~col =
+    if_not_masm (fun () -> D.loc ~file_num ~line ~col ())
+
+  let new_line () = D.new_line ()
+
+  let not_initialized () =
+    Misc.fatal_error "[Asm_directives.initialize] has not been called"
+
+  let define_label label =
+    let lbl_section = Asm_label.section label in
+    let this_section =
+      match !current_section_ref with
+      | None -> not_initialized ()
+      | Some this_section -> this_section
+    in
+    if not (Asm_section.equal lbl_section this_section)
+    then
+      Misc.fatal_errorf
+        "Cannot define label %a intended for section %a in section %a"
+        Asm_label.print label Asm_section.print lbl_section Asm_section.print
+        this_section;
+    (* CR-someday bkhajwal: Be aware of MASM label type annotation *)
+    D.label (Asm_label.encode label)
```

Comment:
I think here we might emit the label into a  wrong section, for example if D.text or D.section was emitted since the last switch_to_section. Are we assuming that all dwarf directives are emitted separately from other assembly?

Reply:
That's right...
I believe we assume that all dwarf directives will be emitted separately from assembly.
For now I'll leave a CR